### PR TITLE
:hammer: gui refactoring

### PR DESCRIPTION
**Pull request by bpetetot**

### CODE REVIEW

**Comment by bpetetot** [comment]
@fabienjuif before going further, I would like that you make a little review of work done : 

First, I have separated store in 2 reducers : 
1. `docgen` reducer contains the whole raw data of docgen api (never updated)
2. `model` reducer contains the properties of the tested component (updated during testing)

Then, I have extracted all docgen utility methods into a single file `redux/utils/docgen.js` (it may be better to manage it otherwise). But all is isolated for now.

I have also added some librairies : `redux-thunk`, `hoc-react-loader` and `reselect`

File: src/gui/components/props/input/input.jsx [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import React from 'react'
 import PropTypes from 'prop-types'
 
-const Input = ({ style, className, name, value, onChange, type }) => {
+const Input = ({ style, className, name, value, onChange, type, input }) => {
```
Comment (fabienjuif) [inline src/gui/components/props/input/input.jsx]: `Input` takes an `input`

what about 
 - letting old `type` (for input) and use a new `propType` (or `rawType`) for the propType
 - or `type` and `inputType`

?

File: src/gui/components/props/input/input.container.js [hunk 2]
```diff
@@ -1,47 +1,52 @@
...
+    value,
+    type: type.name,
+    input: getInputType(type.name),
   }
 }
 
 const mapDispatch = (dispatch, { name }) => ({
-  onChange: type => event => dispatch(setValue(name, getValueFromEvent(type, event))),
+  onChange: (input, type) => event =>
+    dispatch({
```
Comment (fabienjuif) [inline src/gui/components/props/input/input.container.js]: Why don't you use an action creator ?

**Comment by bpetetot** [comment]
@fabienjuif if you dont have other comments, I would like to merge it, and working a bit the style of the gui (in an other PR)... before working on the markdown generation.

File: src/gui/redux/docgen/docgen.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import {
+  REQUEST_DOCGEN,
+  RECEIVED_DOCGEN,
+  RECEIVED_EMPTY_DOCGEN,
+  FAILED_DOCGEN,
+} from './docgen.actions'
+
+export const EMPTY_DOCGEN = {}
+export const LOADING_DOCGEN = {}
```
Comment (fabienjuif) [inline src/gui/redux/docgen/docgen.js]: This kind of use of instance is meh for me.
If you look at state (via `redux-devtools`), you don't know if the state is in an empty state, or a loading state.

This is an unnecessary hack to me.
Why don't we use booleans ?

File: src/gui/redux/docgen/docgen.selectors.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { createSelector } from 'reselect'
+import { LOADING_DOCGEN } from './docgen'
+
+const getDocgen = state => state.docgen
+
+const isLoading = createSelector(getDocgen, docgen => docgen === LOADING_DOCGEN)
```
Comment (fabienjuif) [inline src/gui/redux/docgen/docgen.selectors.js]: hack (https://github.com/Zenika/react-workbench/pull/87/files#r125296633)

**Comment by fabienjuif** [comment]
@bpetetot LGTM, except this idea of references that are clever at first, but lower the debugging and maintainability to me.

File: src/gui/redux/docgen/docgen.js (hunk 3 above)
Comment (bpetetot) [reply src/gui/redux/docgen/docgen.js]: I just liked the way it simplifies reading & writing reducers. 
But I understand your point of view, I change that.

**Review by fabienjuif — APPROVED** [review]
👌

**Pull request merged**